src/praisonai/tests/unit/test_sequential_tool_calling.py

```
@@ -0,0 +1,616 @@
+        result = agent.chat("Analyze my portfolio with Apple, Google, and Microsoft")
+        
+        assert "Portfolio" in result
+        assert mock_completion.call_count == 2
+    
+    @patch('litellm.completion')
+    def test_sequential_tool_retry_on_error(self, mock_completion):
+        """Test that sequential tool calling can retry on transient errors."""
+        # First attempt fails, second succeeds
+        attempt = 0
+        
+        def retry_side_effect(*args, **kwargs):
```

Ruff (0.14.14)

[warning] Unused function argument: `args`

(ARG001)